### PR TITLE
Added explicit specifier to PointerVector ctors

### PR DESCRIPTION
**Pull request by msandre**
This PR proposes to add the `explicit`specifier to the single argument constructors of PointerVector (see #2527). Other classes will be handled in a separate PR. This breaks implicit conversions so implicit conversions are replaced with explicit conversions.

I have checked that the following compiles:
- [x] Core
- [x] SOLID_MECHANICS_APPLICATION
- [x] CONSTITUTIVE_MODELS_APPLICATION
- [x] FLUID_DYNAMICS_APPLICATION
- [x] CONVECTION_DIFFUSION_APPLICATION
- [x] DEM_APPLICATION
- [x] SWIMMING_DEM_APPLICATION
- [x] MESHING_APPLICATION
- [x] EXTERNAL_SOLVERS_APPLICATION
- [x] CONTACT_STRUCTURAL_MECHANICS_APPLICATION
- [x] STRUCTURAL_MECHANICS_APPLICATION
- [x] ALE_APPLICATION
- [x] FSI_APPLICATION
- [x] MAPPING_APPLICATION
- [x] METIS_APPLICATION
- [x] TRILINOS_APPLICATION
- [x] SHAPE_OPTIMIZATION_APPLICATION
- [x] HDF5_APPLICATION

### CODE REVIEW

**Comment by loumalouomega** [comment]
I fixed compilation on processes and tests related with MMG

**Comment by loumalouomega** [comment]
Corrected the ContactStructuralMechanicsApplication

**Comment by loumalouomega** [comment]
@KratosMultiphysics/all

File: applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_condition.cpp [hunk 1]
```diff
@@ -138,12 +138,13 @@ void AugmentedLagrangianMethodFrictionalMortarContactCondition<TDim,TNumNodes,TN
...
         for (IndexType i_geom = 0; i_geom < conditions_points_slave.size(); ++i_geom) {
-            std::vector<PointType::Pointer> points_array (TDim); // The points are stored as local coordinates, we calculate the global coordinates of this points
+            PointerVector< PointType > points_array; // The points are stored as local coordinates, we calculate the global coordinates of this points
+            points_array.reserve(TDim);
             array_1d<BelongType, TDim> belong_array;
             for (IndexType i_node = 0; i_node < TDim; ++i_node) {
                 PointType global_point;
                 slave_geometry.GlobalCoordinates(global_point, conditions_points_slave[i_geom][i_node]);
-                points_array[i_node] = Kratos::make_shared<PointType>(PointType(global_point));
+                points_array.push_back(Kratos::make_shared<PointType>(PointType(global_point)));
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_condition.cpp]: avoid using push_back if you can.

here you know the size, hence define it as 

      PointerVector< PointType > points_array(TDim);

and then do as before but with the ()

      points_array(i_node) = Kratos::make_shared<PointType>(PointType(global_point));

Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_condition.cpp]: Ok

Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_conditions/ALM_frictional_mortar_contact_condition.cpp]: Before i was using [], but it didn't compile, now I see that the operator [] is not the same as (). It is this OK?

**Comment by loumalouomega** [comment]
Solved @RiccardoRossi

File: kratos/tests/sources/test_condition.cpp [hunk 2]
```diff
@@ -41,13 +41,13 @@ namespace Kratos {
             list_nodes[1] = p_node2;
             list_nodes[2] = p_node3;
 
-            auto p_cond = model_part.CreateNewCondition("Condition3D", 1, list_nodes, p_prop);
```
Comment (RiccardoRossi) [inline kratos/tests/sources/test_condition.cpp:44]: well here the cleanest thing would have been to define the list_of_nodes as a PointerVector

**Review by RiccardoRossi — APPROVED** [review]
still let's see if someone has other opinions...so do not hurry too much in merging

**Comment by RiccardoRossi** [comment]
looks like no one else is telling anything ... from my point of view you guys can go ahead

**Pull request merged**